// ngx-order-pipe/esm2022/src/app/order-pipe/ngx-order.module.mjs
/**
 * Created by vadimdez on 20/01/2017.
 */
import { NgModule } from '@angular/core';
import { OrderPipe } from './ngx-order.pipe';
import * as i0 from "@angular/core";
export class OrderModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: OrderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: OrderModule, declarations: [OrderPipe], exports: [OrderPipe] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: OrderModule, providers: [OrderPipe] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: OrderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [OrderPipe],
                    exports: [OrderPipe],
                    providers: [OrderPipe]
                }]
        }] });
